### PR TITLE
easier altcoin integration

### PR DESCRIPTION
**Pull request by oskyk**
With this PR it would be much easier to add another bitcoin clone into btcpy but, but existing functionality is unchanged and no new coins were added.

### CODE REVIEW

File: btcpy/structs/crypto.py [hunk 1]
```diff
@@ -39,14 +40,12 @@ def from_wif(wif, check_network=True):
...
-        if prefix not in {0x80, 0xef}:
+        if prefix not in PrivateKey.wif_prefixes:
             raise ValueError('Unknown private key prefix: {:02x}'.format(prefix))
 
         if check_network:
-            if prefix == 0x80 and not is_mainnet():
-                raise ValueError('Mainnet prefix in testnet environment')
-            elif prefix == 0xef and is_mainnet():
-                raise ValueError('Testnet prefix in mainnet envirnment')
+            if prefix == PrivateKey.wif_prefixes[net_name()]:
```
[the remark this answers is not in the document]
Comment (oskyk) [reply btcpy/structs/crypto.py]: fixed in 7b1a5cf

File: btcpy/setup.py [hunk 2]
```diff
@@ -19,10 +19,11 @@ def setup(network='mainnet', force=False):
     global MAINNET, NETNAME
     if MAINNET is not None and NETNAME != network and not force:
         raise ValueError('Trying to change network type at runtime')
-    if network not in networks:
+    if network not in NETWORKS:
         raise ValueError('Unknown network type: {}'.format(network))
     MAINNET = (network == 'mainnet')
```
[the remark this answers is not in the document]
Comment (oskyk) [reply btcpy/setup.py]: is_mainet() is only used in segwit addresses now (no other clone supports segwit). In other parts of code network is determined by net_name() which will work correctly even with mainnet_dash or similar

**Comment by SimoneBronzini** [comment]
Great job! I will take some time to thoroughly review it and run the appropriate tests.

File: btcpy/structs/address.py [hunk 3]
```diff
@@ -50,9 +50,10 @@ def get_codec():
...
 
     def __init__(self, addr_type, hashed_data, mainnet=None):
-        if mainnet is None:
-            mainnet = is_mainnet()
-        network = 'mainnet' if mainnet else 'testnet'
+        network = 'mainnet' if mainnet is True else mainnet
+        network = 'testnet' if mainnet is True else network
+        if network is None:
+            network = net_name()
         self.network = network
```
[the remark this answers is not in the document]
Comment (SimoneBronzini) [reply btcpy/structs/address.py]: Yeah, we might consider breaking it if the change is important enough. I'll have a stronger opinion about this as soon as I review the whole thing.

[the remark this answers is not in the document]
Comment (oskyk) [reply btcpy/structs/address.py]: thanks for highlighting there is a bug fixed in 2131472

File: btcpy/setup.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SimoneBronzini) [reply btcpy/setup.py]: Actually, Litecoin supports SegWit.

[the remark this answers is not in the document]
Comment (oskyk) [reply btcpy/setup.py]: Sewgit no longer use is_mainnet() and I can't find any other usages of this function should I remove it?

File: btcpy/structs/address.py [hunk 4]
```diff
@@ -50,9 +50,10 @@ def get_codec():
         return Base58Codec
 
     def __init__(self, addr_type, hashed_data, mainnet=None):
-        if mainnet is None:
-            mainnet = is_mainnet()
-        network = 'mainnet' if mainnet else 'testnet'
+        network = 'mainnet' if mainnet is True else None
+        network = 'testnet' if mainnet is False else network
```
Comment (SimoneBronzini) [inline btcpy/structs/address.py]: Readability might be improved here by explicitly checking the three cases (`mainnet is True`, `mainnet is False`, `mainnet is None`)

File: btcpy/constants.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+from btcpy.lib.codecs import Base58Codec, Bech32Codec
```
Comment (SimoneBronzini) [inline btcpy/constants.py]: This looks like a circular import (this imports `codecs`, `codecs` imports `setup`, `setup` imports `constants`). Please double-check with unit tests.

File: btcpy/constants.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+from btcpy.structs.crypto import PrivateKey
```
Comment (SimoneBronzini) [inline btcpy/constants.py]: Same as above.

**Comment by SimoneBronzini** [comment]
I did a first review of the code, summarised in the comments above. Please double-check with unit tests. In the meantime, new commits created a merge conflict but I think the only problem is in `crypto.py`, a typo fix on line 49, so it should be easy to adapt.

File: btcpy/constants.py (hunk 5 above)
Comment (oskyk) [reply btcpy/constants.py]: if it would be unittest would fail on unit.py:26 but the unit test result is same as on master

**Comment by oskyk** [comment]
Confilict is fixed, integration test is passing, unittest results are same as on master

**Comment by SimoneBronzini** [comment]
My tests are failing because of circular imports, are you sure your tests are running the updated version of `btcpy` and not an old version installed at system level? I added a note [here](https://github.com/chainside/btcpy#contributing-and-running-tests) on how to make sure one is running the version currently in the repo during testing.

**Comment by oskyk** [comment]
Sorry about that. It's fixed now

**Comment by SimoneBronzini** [comment]
Sorry for the very late feedback, very busy period. Actually, to fix the circular import problem it would have been enough to move the import statements inside the `init_constants` method. I think that solution would be better instead of having every module directly access the global dictionary.

**Comment by peerchemist** [comment]
Preferably the lib should take argument in form of NamedTuple, or inherit a class carrying custom parameters for *coin. In this way upstream won't carry the settings in it's repo (except for Bitcoin) but each individual application will provide custom parameters upon initialization.

**Comment by SimoneBronzini** [comment]
@peerchemist seems like a good idea, do you think you can do a PR (or modify this one) with the change you have in mind?

**Comment by peerchemist** [comment]
I can try, I can't promise on the time-frame though as I am horribly busy lately.

**Comment by peerchemist** [comment]
I've submitted my take on this problem here: https://github.com/chainside/btcpy/pull/25

As you can now custom (application defined) Constants class can be passed to library upon setup.
Also, now setup.py defines the network parameters for the library which makes more sense than it made before.

**Pull request closed without merging**